lib/puppet/transaction.rb

```diff
@@ -287,15 +291,21 @@ def failed_dependencies?(resource)
     # explosion of edges, we also ended up reporting failures for containers
     # like class and stage.  This is undesirable; while just skipping the
     # output isn't perfect, it is RC-safe. --daniel 2011-06-07
-    suppress_report = (resource.class == Puppet::Type.type(:whit))
-
+    class_or_stage = resource.class == Puppet::Type.type(:whit)
+    # With coalesce_class_failures, notify about class dependency failures ... just once per class. TJK 2019-09-09
     s = resource_status(resource)
     if s && s.dependency_failed?
-      # See above. --daniel 2011-06-06
-      unless suppress_report then
-        s.failed_dependencies.find_all { |d| !(@failed_dependencies_already_notified.include?(d.ref)) }.each do |dep|
-          resource.notice _("Dependency %{dep} has failures: %{status}") % { dep: dep, status: resource_status(dep).failed }
-          @failed_dependencies_already_notified.add(dep.ref)
+      if @coalesce_class_failures == true && class_or_stage == true then
+        s.failed_dependencies.find_all { |d| !(@failed_class_dependencies_already_notified.include?(d.ref)) }.each do |dep|
```

Comment:
Can we just calculate the difference like?

```ruby
(s.failed_dependencies.map(&:ref) - @failed_class_dependencies_already_notified).each do |dep|
```

I'm not sure if that's faster than what you have or how it affects the order that the `notice` messages are printed.

Reply:
Yes, but ...

This results in a different form of code than in the `else` (which is the current code) and I had to convert to the `@failed_class_dependencies_already_notified)` Set to an Array, as `Set.difference` did not work as expected.